File: backend/internal/repository/project.go
/*
 * @Author: shanlonglong danlonglong@weimiao.cn
 * @Date: 2025-02-21 14:54:42
 * @LastEditors: shanlonglong danlonglong@weimiao.cn
 * @LastEditTime: 2025-02-24 17:48:33
 * @FilePath: \go_react_erp\backend\internal\repository\project.go
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
package repository

import (
	"erp-sys/internal/model"
	"erp-sys/pkg/database"
	"time"
)

type ProjectRepository struct{}

func NewProjectRepository() *ProjectRepository {
	return &ProjectRepository{}
}

// GetCurrentProject 获取当前项目
func (r *ProjectRepository) GetCurrentProject() (*model.Project, error) {
	var project model.Project
	err := database.DB.Preload("Tasks").First(&project).Error
	if err != nil {
		return nil, err
	}
	return &project, nil
}

// UpdateTaskPosition 更新任务位置和开始时间
func (r *ProjectRepository) UpdateTaskPosition(taskID uint, phase model.Phase, newStartDate time.Time) error {
	// 先获取原任务信息
	var task model.Task
	if err := database.DB.First(&task, taskID).Error; err != nil {
		return err
	}
	
	// 计算任务持续时间
	duration := task.EndTime.Sub(task.StartTime)
	
	// 只更新需要修改的字段
	updates := map[string]interface{}{
		"start_time": newStartDate,
		"end_time":   newStartDate.Add(duration),
		"phase":      phase,
	}
	
	// 使用 Updates 方法只更新指定字段
	return database.DB.Model(&task).Updates(updates).Error
}

// UpdateTaskStatus 更新任务状态
func (r *ProjectRepository) UpdateTaskStatus(taskID uint, status model.TaskStatus) error {
	return database.DB.Model(&model.Task{}).
		Where("id = ?", taskID).
		Update("status", status).Error
}

// CreateTask 创建任务
func (r *ProjectRepository) CreateTask(task *model.Task) error {
	return database.DB.Create(task).Error
} 

File: sys-admin/src/services/project.ts
/*
 * @Author: shanlonglong danlonglong@weimiao.cn
 * @Date: 2025-02-21 14:37:51
 * @LastEditors: shanlonglong danlonglong@weimiao.cn
 * @LastEditTime: 2025-02-21 16:40:37
 * @FilePath: \go_react_erp\sys-admin\src\services\project.ts
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import request from '@/utils/request';
import type { Project, TaskStatus } from '@/types/project';

// 获取项目详情
export const getProject = (id?: string) => {
  return request.get<{ data: Project }>('/projects/current');
};

// 更新任务位置
export const updateTaskPosition = (params: {
  taskId: number;
  newPhase: string;
  newStartDate: string;
}) => {
  return request(`/tasks/${params.taskId}/position`, {
    method: 'PUT',
    data: {
      newPhase: params.newPhase,
      newStartDate: params.newStartDate,
    },
  });
};

// 更新任务状态
export const updateTaskStatus = (taskId: string, status: TaskStatus) => {
  return request.put(`/tasks/${taskId}/status`, { status });
};

// 创建任务
export const createTask = (data: {
  projectId: string;
  title: string;
  description: string;
  startTime: string;
  endTime: string;
  phase: string;
}) => {
  return request.post('/tasks', data);
};

// 更新任务
export const updateTask = (taskId: string, data: {
  title?: string;
  description?: string;
  startTime?: string;
  endTime?: string;
}) => {
  return request.put(`/tasks/${taskId}`, data);
};

// 删除任务
export const deleteTask = (taskId: string) => {
  return request.delete(`/tasks/${taskId}`);
};

// 获取项目列表
export const getProjectList = (params: {
  current?: number;
  pageSize?: number;
  name?: string;
}) => {
  return request.get('/projects', { params });
};

// 创建项目
export const createProject = (data: {
  name: string;
  startDate: string;
  endDate: string;
  description?: string;
}) => {
  return request.post('/projects', data);
};

// 更新项目
export const updateProject = (id: string, data: {
  name?: string;
  startDate?: string;
  endDate?: string;
  description?: string;
}) => {
  return request.put(`/projects/${id}`, data);
};

// 删除项目
export const deleteProject = (id: string) => {
  return request.delete(`/projects/${id}`);
}; 